File: 0809. Expressive Words/Solution.py
from typing import List


class Solution:
    def expressiveWords(self, s: str, words: List[str]) -> int:
        def isExpressive(s: str, t: str) -> bool:
            sn, tn = len(s), len(t)
            si = ti = 0
            while si < sn and ti < tn:
                if s[si] != t[ti]:
                    return False
                sj = si + 1
                while sj < sn and s[sj] == s[si]:
                    sj += 1
                tj = ti + 1
                while tj < tn and t[tj] == t[ti]:
                    tj += 1

                s_cnt, t_cnt = sj - si, tj - ti
                if (t_cnt < s_cnt and s_cnt < 3) or s_cnt < t_cnt:
                    return False
                si, ti = sj, tj
            return si == sn and ti == tn

        return sum(isExpressive(s, word) for word in words)

File: 0809. Expressive Words/Solution.cpp
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    int expressiveWords(string s, vector<string>& words)
    {
        int res = 0;
        for (const string& word : words)
            res += isExpressive(s, word);
        return res;
    }

private:
    bool isExpressive(const string& s, const string& t)
    {
        int si = 0, ti = 0, sn = s.size(), tn = t.size();
        while (si < sn && ti < tn) {
            if (s[si] != t[ti])
                return false;
            int sj = si + 1;
            while (sj < sn && s[sj] == s[si])
                ++sj;

            int tj = ti + 1;
            while (tj < tn && t[tj] == t[ti])
                ++tj;

            int s_cnt = sj - si, t_cnt = tj - ti;
            if ((t_cnt < s_cnt && s_cnt < 3) || s_cnt < t_cnt)
                return false;
            si = sj;
            ti = tj;
        }
        return si == sn && ti == tn;
    }
};
